src/decoder/lattice-faster-decoder-combine.cc

```diff
@@ -0,0 +1,1092 @@
+  for (IterType iter = cur_toks_.begin(); iter != cur_toks_.end(); iter++) {
+    cur_queue.push_back(iter->first);
+    iter->second->in_current_queue = true;
+  }
+
+  // Iterator the "cur_queue" to process non-emittion and emittion arcs in fst.
+  while (!cur_queue.empty()) {
+    StateId state = cur_queue.back();
+    cur_queue.pop_back();
+
+    KALDI_ASSERT(cur_toks_.find(state) != cur_toks_.end());
+    Token *tok = cur_toks_[state];
+    BaseFloat cur_cost = tok->tot_cost;
+    if (cur_cost > cur_cutoff)  // Don't bother processing successors.
+      continue;
+    // If "tok" has any existing forward links, delete them,
+    // because we're about to regenerate them.  This is a kind
+    // of non-optimality (remember, this is the simple decoder),
+    DeleteForwardLinks(tok);  // necessary when re-visiting
+    tok->links = NULL;
+    for (fst::ArcIterator<FST> aiter(*fst_, state);
+         !aiter.Done();
+         aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      bool changed;
+      if (arc.ilabel == 0) {  // propagate nonemitting
+        BaseFloat graph_cost = arc.weight.Value();
+        BaseFloat tot_cost = cur_cost + graph_cost;
+        if (tot_cost < cur_cutoff) {
+          Token *new_tok = FindOrAddToken(arc.nextstate, frame, tot_cost,
+                                          tok, &cur_toks_, &changed);
+
+          // Add ForwardLink from tok to new_tok. Put it on the head of
+          // tok->link list
+          tok->links = new ForwardLinkT(new_tok, 0, arc.olabel,
+                                        graph_cost, 0, tok->links);
+          
+          // "changed" tells us whether the new token has a different
+          // cost from before, or is new.
+          if (changed && !new_tok->in_current_queue) {
+            cur_queue.push_back(arc.nextstate);
+            new_tok->in_current_queue = true;
+          }
+        }
+      } else {  // propagate emitting
+        BaseFloat graph_cost = arc.weight.Value(),
+                  ac_cost = cost_offset - decodable->LogLikelihood(frame, arc.ilabel),
+                  cur_cost = tok->tot_cost,
+                  tot_cost = cur_cost + ac_cost + graph_cost;
+        if (tot_cost > next_cutoff) continue;
+        else if (tot_cost + adaptive_beam < next_cutoff)
+          next_cutoff = tot_cost + adaptive_beam;  // a tighter boundary for emitting
+        
+        // no change flag is needed
+        Token *next_tok = FindOrAddToken(arc.nextstate, frame + 1, tot_cost,
+                                         tok, &next_toks_, NULL);
+        // Add ForwardLink from tok to next_tok. Put it on the head of tok->link
+        // list
+        tok->links = new ForwardLinkT(next_tok, arc.ilabel, arc.olabel,
+                                      graph_cost, ac_cost, tok->links);
+      }
+    }  // for all arcs
+    tok->in_current_queue = false;  // out of queue
+  }  // end of while loop
+  KALDI_VLOG(6) << "toks after: " << cur_toks_.size();
```

Comment:
KALDI_VLOG(6) << "Number of tokens active on frame " << NumFramesDecoded()-1
                  << " is " << cur_toks_.size();